src/auth/models.cpp

```diff
@@ -186,112 +203,102 @@ bool operator==(const Permissions &first, const Permissions &second) {
 
 bool operator!=(const Permissions &first, const Permissions &second) { return !(first == second); }
 
-const std::string ASTERISK = "*";
-
-FineGrainedAccessPermissions::FineGrainedAccessPermissions(const std::unordered_set<std::string> &grants,
-                                                           const std::unordered_set<std::string> &denies)
-    : grants_(grants), denies_(denies) {}
+FineGrainedAccessPermissions::FineGrainedAccessPermissions(const std::unordered_map<std::string, uint64_t> &permissions,
+                                                           const std::optional<uint64_t> &global_permission)
+    : permissions_(permissions), global_permission_(global_permission) {}
 
-PermissionLevel FineGrainedAccessPermissions::Has(const std::string &permission) const {
-  if ((denies_.size() == 1 && denies_.find(ASTERISK) != denies_.end()) || denies_.find(permission) != denies_.end()) {
-    return PermissionLevel::DENY;
-  }
+PermissionLevel FineGrainedAccessPermissions::Has(const std::string &permission,
+                                                  const LabelPermission label_permission) {
+  uint64_t concrete_permission = permissions_.contains(permission)
+                                     ? permissions_[permission]
+                                     : (global_permission_.has_value() ? global_permission_.value() : 0);
```

Comment:
I think this is more readable in my opinion, multiple ternary operators are a bit confusing, definitely include `<functional>` for invoke
```suggestion
  const auto concrete_permission = std::invoke([&]() -> uint64_t {
    if (permissions_.contains(permission)) {
      return permissions_[permission];
    }
    if (global_permission_.has_value()) {
      return global_permission_.value();
    }
    return 0;
  });
```